File: typescript/src/agents/research/agent.ts
/**
 * `ResearchAgent` — strict-output ReAct loop with per-snippet compression.
 *
 * Mirrors the Python implementation in `compresr/agents/research/agent.py`.
 * Bypasses `_Engine.run` so we can set `tool_choice` per step and compress
 * each tool result before it enters the conversation.
 */

import { parseResearchOutput } from './parser.js';
import { DEFAULT_RESEARCH_SYSTEM_PROMPT } from './prompts.js';
import type { Citation, ResearchResult, ResearchUsage, Step } from './types.js';

interface CompresrClientLike {
  compress(opts: {
    context: string;
    query?: string;
    compressionModelName?: string;
  }): Promise<{
    data?: { compressedContext?: string; compressed_context?: string };
  }>;
}

export interface ChatModelLike {
  bindTools(
    tools: ReadonlyArray<unknown>,
    options: { tool_choice: string }
  ): { invoke(messages: unknown[]): Promise<unknown> };
}

export interface SearchToolLike {
  name?: string;
  invoke(args: Record<string, unknown>): Promise<unknown>;
}

export interface ResearchEngine {
  readonly provider: string;
  readonly defaultModelName?: string;
  _resolveModel(model?: string): string;
  _getChatModel(modelName: string): Promise<ChatModelLike>;
  readonly _compresrClient: CompresrClientLike;
  readonly _promptCacheConfig: {
    readonly enabled: boolean;
    readonly ttl: '5m' | '1h';
    readonly minMessages: number;
  };
}

interface AIMessageLike {
  content?: unknown;
  tool_calls?: ReadonlyArray<{
    id?: string;
    name?: string;
    args?: Record<string, unknown>;
  }>;
  usage_metadata?: Record<string, unknown>;
}

const DEFAULT_MAX_CONTEXT_TOKENS = 120_000;
const DEFAULT_MAX_STEPS = 10;
const DEFAULT_MIN_COMPRESS_TOKENS = 100;

const CACHE_READ_ALIASES = [
  'cached_tokens',
  'cache_read',
  'priority_cache_read',
  'flex_cache_read',
  'cached_content_token_count',
] as const;

export interface ResearchAgentOptions {
  engine: ResearchEngine;
  searchTool: SearchToolLike;
  maxSteps?: number;
  systemPrompt?: string;
  compressSnippets?: boolean;
  compressionModel?: string;
  minCompressTokens?: number;
  maxContextTokens?: number;
}

export interface ResearchRunOptions {
  model?: string;
}

/**
 * Internal result of {@link ResearchAgent._runLoop}.
 *
 * Exposed to the engine's solo-WebSearchTool fast-path so it can feed the
 * raw message chain to {@link CompresrEngine._normalize}, while {@link ResearchAgent.run}
 * keeps producing a typed {@link ResearchResult}.
 */
export interface ResearchLoopState {
  messages: unknown[];
  aiHistory: AIMessageLike[];
  trajectory: Step[];
  lastAi: AIMessageLike | null;
}

function estimateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

function stringifyContent(content: unknown): string {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) {
    const parts: string[] = [];
    for (const block of content) {
      if (block && typeof block === 'object' && 'text' in block) {
        const t = (block as { text?: unknown }).text;
        if (typeof t === 'string') parts.push(t);
      } else if (typeof block === 'string') {
        parts.push(block);
      }
    }
    return parts.join('');
  }
  if (content == null) return '';
  if (typeof content === 'number' || typeof content === 'boolean') return String(content);
  return JSON.stringify(content);
}

function readCacheTokens(
  details: Record<string, unknown> | undefined
): { read: number; create: number } {
  if (!details) return { read: 0, create: 0 };
  let read = 0;
  for (const alias of CACHE_READ_ALIASES) {
    const v = details[alias];
    if (typeof v === 'number') read += v;
  }
  // langchain-anthropic zeros ``cache_creation`` when ttl-specific keys are set.
  const ttl5 = typeof details['ephemeral_5m_input_tokens'] === 'number'
    ? (details['ephemeral_5m_input_tokens']) : 0;
  const ttl1 = typeof details['ephemeral_1h_input_tokens'] === 'number'
    ? (details['ephemeral_1h_input_tokens']) : 0;
  const gen = typeof details['cache_creation'] === 'number'
    ? (details['cache_creation']) : 0;
  const create = ttl5 + ttl1 > 0 ? ttl5 + ttl1 : gen;
  return { read, create };
}

export class ResearchAgent {
  private readonly engine: ResearchEngine;
  private readonly searchTool: SearchToolLike;
  private readonly maxSteps: number;
  private readonly systemPrompt: string;
  private readonly compressSnippets: boolean;
  private readonly compressionModel: string;
  private readonly minCompressTokens: number;
  private readonly maxContextTokens: number;

  constructor(options: ResearchAgentOptions) {
    this.engine = options.engine;
    this.searchTool = options.searchTool;
    this.maxSteps = Math.max(1, options.maxSteps ?? DEFAULT_MAX_STEPS);
    this.systemPrompt = options.systemPrompt ?? DEFAULT_RESEARCH_SYSTEM_PROMPT;
    this.compressSnippets = options.compressSnippets ?? true;
    this.compressionModel = options.compressionModel ?? 'latte_v1';
    this.minCompressTokens = options.minCompressTokens ?? DEFAULT_MIN_COMPRESS_TOKENS;
    this.maxContextTokens = options.maxContextTokens ?? DEFAULT_MAX_CONTEXT_TOKENS;
  }

  async run(question: string, runOptions: ResearchRunOptions = {}): Promise<ResearchResult> {
    const state = await this._runLoop(question, runOptions);
    const searchCalls = state.trajectory.filter((s) => s.type === 'search').length;
    const usage = this.aggregateUsage(state.aiHistory, searchCalls);
    const finalText = state.lastAi ? stringifyContent(state.lastAi.content) : '';
    const parsed = parseResearchOutput(finalText);
    const citations = this.collectCitations(parsed.citationUrls, state.trajectory);

    return {
      answer: parsed.answer,
      explanation: parsed.explanation,
      confidence: parsed.confidence,
      text: finalText,
      citations,
      trajectory: state.trajectory,
      usage,
      raw: state.lastAi,
    };
  }

  /**
   * Execute the strict-output ReAct loop and return raw state.
   *
   * Shared by {@link run} (which parses ``ResearchResult``) and the engine's
   * solo-WebSearchTool fast-path (which feeds ``state.messages`` into
   * the engine's normalizer).
   *
   * @internal — public for the engine's fast-path; not a stable surface.
   */
  async _runLoop(
    question: string,
    runOptions: ResearchRunOptions = {}
  ): Promise<ResearchLoopState> {
    const { SystemMessage, HumanMessage, ToolMessage } = await import(
      '@langchain/core/messages'
    );

    const modelName = this.engine._resolveModel(runOptions.model);
    const chat = await this.engine._getChatModel(modelName);

    const messages: unknown[] = [
      new SystemMessage(this.systemPrompt),
      new HumanMessage(question),
    ];
    const aiHistory: AIMessageLike[] = [];
    const trajectory: Step[] = [];
    let lastAi: AIMessageLike | null = null;

    for (let stepIdx = 0; stepIdx < this.maxSteps; stepIdx++) {
      this.maybeTruncate(messages);
      const toolChoice = stepIdx === this.maxSteps - 1 ? 'none' : 'auto';
      const bound = chat.bindTools([this.searchTool], { tool_choice: toolChoice });
      const invokeMessages = this.applyCacheControl(messages);

      const t0 = Date.now();
      let response: AIMessageLike;
      try {
        response = (await bound.invoke(invokeMessages)) as AIMessageLike;
      } catch (exc) {
        trajectory.push({
          type: 'error',
          text: exc instanceof Error ? exc.message : String(exc),
          latencyS: (Date.now() - t0) / 1000,
        });
        break;
      }
      const stepLatencyS = (Date.now() - t0) / 1000;

      aiHistory.push(response);
      lastAi = response;
      messages.push(response);

      const toolCalls = response.tool_calls ?? [];

      if (toolCalls.length === 0) {
        trajectory.push({
          type: 'answer',
          text: stringifyContent(response.content),
          latencyS: stepLatencyS,
        });
        break;
      }

      for (const tc of toolCalls) {
        const args = tc.args ?? {};
        const query = typeof args['query'] === 'string' ? (args['query']) : undefined;
        trajectory.push({ type: 'search', query, latencyS: stepLatencyS });

        const rawResult = await this.invokeSearch(args);
        const finalResult = this.compressSnippets
          ? await this.compress(rawResult, query ?? question)
          : rawResult;
        trajectory.push({ type: 'tool_result', chars: finalResult.length, text: finalResult });

        messages.push(
          new ToolMessage({
            content: finalResult,
            tool_call_id: tc.id ?? '',
            name: this.searchTool.name ?? 'search_web',
          })
        );
      }
    }

    return { messages, aiHistory, trajectory, lastAi };
  }

  /**
   * Mirrors AnthropicPromptCachingMiddleware for the bypass path — research
   * doesn't go through createAgent so we stamp cache_control manually.
   */
  private applyCacheControl(messages: unknown[]): unknown[] {
    if (this.engine.provider !== 'anthropic') return messages;
    const cfg = this.engine._promptCacheConfig;
    if (!cfg?.enabled) return messages;
    if (messages.length < cfg.minMessages) return messages;

    const marker = { type: 'ephemeral', ttl: cfg.ttl };
    const last = messages[messages.length - 1] as {
      content?: unknown;
      constructor?: new (args: unknown) => unknown;
    };
    const content = last?.content;
    let newContent: unknown[];
    if (typeof content === 'string') {
      newContent = [{ type: 'text', text: content, cache_control: marker }];
    } else if (Array.isArray(content) && content.length > 0) {
      newContent = [...(content as unknown[])];
      const tail = newContent[newContent.length - 1];
      if (tail !== null && typeof tail === 'object') {
        newContent[newContent.length - 1] = {
          ...(tail as Record<string, unknown>),
          cache_control: marker,
        };
      } else {
        const tailText = typeof tail === 'string' ? tail : '';
        newContent.push({ type: 'text', text: tailText, cache_control: marker });
      }
    } else {
      return messages;
    }
    const Ctor = last.constructor as new (arg: { content: unknown[] }) => unknown;
    const patched = new Ctor({ content: newContent });
    return [...messages.slice(0, -1), patched];
  }

  private async invokeSearch(args: Record<string, unknown>): Promise<string> {
    try {
      const out = await this.searchTool.invoke(args);
      if (typeof out === 'string') return out;
      if (out == null) return '';
      return JSON.stringify(out);
    } catch (exc) {
      const msg = exc instanceof Error ? exc.message : String(exc);
      // eslint-disable-next-line no-console
      console.warn(`[compresr.research] search tool raised: ${msg}; continuing.`);
      return `Search error: ${msg}`;
    }
  }

  private async compress(text: string, query: string): Promise<string> {
    if (!text || estimateTokens(text) < this.minCompressTokens) return text;
    const client = this.engine._compresrClient;
    if (!client?.compress) return text;
    try {
      const resp = await client.compress({
        context: text,
        query,
        compressionModelName: this.compressionModel,
      });
      const cc = resp?.data?.compressedContext ?? resp?.data?.compressed_context;
      return typeof cc === 'string' && cc.length > 0 ? cc : text;
    } catch {
      return text;
    }
  }

  private maybeTruncate(messages: unknown[]): void {
    if (!this.maxContextTokens) return;
    const total = () =>
      messages.reduce<number>(
        (acc, m) => acc + estimateTokens(stringifyContent((m as { content?: unknown })?.content)),
        0
      );
    let cur = total();
    if (cur <= this.maxContextTokens) return;
    let guard = 0;
    while (cur > this.maxContextTokens && messages.length >= 4 && guard < 4) {
      messages.splice(2, 2);
      cur = total();
      guard++;
    }
  }

  private aggregateUsage(aiMessages: AIMessageLike[], searchCalls = 0): ResearchUsage {
    const totals: Record<string, number> = {
      input_tokens: 0,
      output_tokens: 0,
      cache_read_tokens: 0,
      cache_creation_tokens: 0,
    };
    for (const m of aiMessages) {
      const um = m?.usage_metadata ?? {};
      const inTok = typeof um['input_tokens'] === 'number' ? (um['input_tokens']) : 0;
      totals['output_tokens'] += typeof um['output_tokens'] === 'number'
        ? (um['output_tokens']) : 0;

      const details = um['input_token_details'] as Record<string, unknown> | undefined;
      const { read, create } = readCacheTokens(details);
      totals['cache_read_tokens'] += read;
      totals['cache_creation_tokens'] += create;

      // langchain-anthropic 1.x inflates input_tokens to fresh+read+create.
      const overcount = read + create;
      totals['input_tokens'] += Math.max(0, inTok - overcount);
    }
    return {
      input_tokens: totals['input_tokens'],
      output_tokens: totals['output_tokens'],
      cache_read_tokens: totals['cache_read_tokens'],
      cache_creation_tokens: totals['cache_creation_tokens'],
      calls: aiMessages.length,
      search_calls: searchCalls,
    };
  }

  private collectCitations(parsedUrls: ReadonlyArray<string>, trajectory: Step[]): Citation[] {
    const urls: string[] = [];
    for (const u of parsedUrls) if (!urls.includes(u)) urls.push(u);
    const urlRe = /https?:\/\/[^\s,;<>"')]+/g;
    for (const step of trajectory) {
      if (step.type !== 'tool_result' || !step.text) continue;
      let m: RegExpExecArray | null;
      while ((m = urlRe.exec(step.text)) !== null) {
        if (!urls.includes(m[0])) urls.push(m[0]);
      }
    }
    return urls.map((url) => ({ url }));
  }
}

File: python/compresr/agents/research/agent.py
"""``ResearchAgent`` — strict-output ReAct loop with per-snippet compression.

Bypasses ``_Engine.run`` so we can control ``tool_choice`` per step (forces
the model to commit on the final step) and compress each tool result against
the live query before it enters the conversation. Cache control is stamped
manually for Anthropic since ``create_agent`` is not in the call path.
"""

from __future__ import annotations

import logging
import re
import time
from dataclasses import dataclass, field
from typing import Any, Optional, Sequence

from .parser import parse_research_output
from .prompts import DEFAULT_RESEARCH_SYSTEM_PROMPT
from .types import Citation, ResearchResult, ResearchUsage, Step

logger = logging.getLogger(__name__)

_DEFAULT_MAX_CONTEXT_TOKENS = 120_000
_DEFAULT_MAX_STEPS = 10
_DEFAULT_MIN_COMPRESS_TOKENS = 100


def _estimate_tokens(text: str) -> int:
    return max(1, len(text) // 4)


@dataclass
class _LoopState:
    """Internal result of :meth:`ResearchAgent._run_loop`.

    Exposed to the engine's solo-WebSearchTool fast-path so it can feed the
    raw message chain to :meth:`_Engine._normalize` while the public
    :meth:`run` keeps producing a typed :class:`ResearchResult`.
    """

    messages: list
    ai_history: list = field(default_factory=list)
    trajectory: list = field(default_factory=list)
    last_ai: Optional[Any] = None


class ResearchAgent:
    """Multi-step web-research agent with on-the-fly compression.

    Args:
        engine: A live :class:`compresr.agents.engine._Engine`.
        search_tool: LangChain ``BaseTool`` taking ``query: str``.
        max_steps: Hard cap on LLM turns. On the final turn ``tool_choice``
            is set to ``"none"`` so the model has to commit.
        system_prompt: Override the default research prompt.
        compress_snippets: Compress each tool result via
            ``client.compress(query=live_query)`` before appending.
        compression_model: Compresr model name (default ``"latte_v1"``).
        min_compress_tokens: Skip compression for tool results below this size.
        max_context_tokens: Drop the oldest tool-result pair when context
            exceeds this cap.
    """

    def __init__(
        self,
        *,
        engine: Any,
        search_tool: Any,
        max_steps: int = _DEFAULT_MAX_STEPS,
        system_prompt: Optional[str] = None,
        compress_snippets: bool = True,
        compression_model: str = "latte_v1",
        min_compress_tokens: int = _DEFAULT_MIN_COMPRESS_TOKENS,
        max_context_tokens: int = _DEFAULT_MAX_CONTEXT_TOKENS,
    ) -> None:
        self._engine = engine
        self._search_tool = search_tool
        self._max_steps = max(1, int(max_steps))
        self._system_prompt = system_prompt or DEFAULT_RESEARCH_SYSTEM_PROMPT
        self._compress_snippets = bool(compress_snippets)
        self._compression_model = compression_model
        self._min_compress_tokens = int(min_compress_tokens)
        self._max_context_tokens = int(max_context_tokens)

    def run(self, question: str, *, model: Optional[str] = None) -> ResearchResult:
        state = self._run_loop(question, model=model)
        usage = self._aggregate_usage(
            state.ai_history,
            search_calls=sum(1 for s in state.trajectory if s.type == "search"),
        )
        final_text = self._stringify(state.last_ai.content) if state.last_ai is not None else ""
        parsed = parse_research_output(final_text)
        citations = self._collect_citations(parsed.citation_urls, state.trajectory)

        return ResearchResult(
            answer=parsed.answer,
            explanation=parsed.explanation,
            confidence=parsed.confidence,
            text=final_text,
            citations=citations,
            trajectory=state.trajectory,
            usage=usage,
            raw=state.last_ai,
        )

    def _run_loop(
        self,
        question: str,
        *,
        model: Optional[str] = None,
        extra_chat_kwargs: Optional[dict] = None,
    ) -> _LoopState:
        """Execute the strict-output ReAct loop and return raw state.

        Shared by :meth:`run` (which parses ``ResearchResult``) and the
        engine's solo-WebSearchTool fast-path (which feeds ``state.messages``
        into :meth:`_Engine._normalize`).

        ``extra_chat_kwargs`` lets the engine path bake per-call LLM knobs
        (``max_tokens``, ``temperature``…) into the cached chat model so
        they survive ``bind_tools``.
        """
        from langchain_core.messages import (  # type: ignore[import-not-found]
            HumanMessage,
            SystemMessage,
            ToolMessage,
        )

        effective_model = self._engine._resolve_model(model)
        chat = self._engine._get_or_build_chat(effective_model, extra_kwargs=extra_chat_kwargs)

        messages: list = [
            SystemMessage(content=self._system_prompt),
            HumanMessage(content=question),
        ]
        ai_history: list = []
        trajectory: list[Step] = []
        last_ai: Optional[Any] = None

        for step_idx in range(self._max_steps):
            self._maybe_truncate(messages)

            tool_choice = "none" if step_idx == self._max_steps - 1 else "auto"
            bound = chat.bind_tools([self._search_tool], tool_choice=tool_choice)
            invoke_messages = self._apply_cache_control(messages)

            t0 = time.time()
            try:
                response = bound.invoke(invoke_messages)
            except Exception as exc:  # noqa: BLE001
                trajectory.append(
                    Step(type="error", text=str(exc), latency_s=round(time.time() - t0, 3))
                )
                break
            step_latency = round(time.time() - t0, 3)

            ai_history.append(response)
            last_ai = response
            messages.append(response)

            tool_calls = getattr(response, "tool_calls", None) or []
            if not tool_calls:
                trajectory.append(
                    Step(
                        type="answer",
                        text=self._stringify(response.content),
                        latency_s=step_latency,
                    )
                )
                break

            for tc in tool_calls:
                args = tc.get("args") if isinstance(tc, dict) else {}
                args = args if isinstance(args, dict) else {}
                query = args.get("query")
                trajectory.append(Step(type="search", query=query, latency_s=step_latency))

                raw_result = self._invoke_search(args)
                final_result = (
                    self._compress(raw_result, query or question)
                    if self._compress_snippets
                    else raw_result
                )
                trajectory.append(
                    Step(
                        type="tool_result",
                        chars=len(final_result),
                        text=final_result,
                    )
                )

                tool_call_id = tc.get("id") if isinstance(tc, dict) else None
                messages.append(
                    ToolMessage(
                        content=final_result,
                        tool_call_id=tool_call_id or "",
                        name=getattr(self._search_tool, "name", "search_web"),
                    )
                )

        return _LoopState(
            messages=messages,
            ai_history=ai_history,
            trajectory=trajectory,
            last_ai=last_ai,
        )

    def _apply_cache_control(self, messages: list) -> list:
        """Stamp Anthropic ephemeral cache_control on the last message.

        Mirrors ``AnthropicPromptCachingMiddleware`` for the bypass path —
        the research agent does not go through ``create_agent`` so the
        engine's middleware never runs. No-op for non-Anthropic providers.
        """
        provider = getattr(self._engine, "_provider", None)
        if provider != "anthropic":
            return messages
        if not getattr(self._engine, "_enable_prompt_cache", False):
            return messages
        min_msgs = getattr(self._engine, "_prompt_cache_min_messages", 2)
        if len(messages) < min_msgs:
            return messages

        ttl = getattr(self._engine, "_prompt_cache_ttl", "5m")
        marker = {"type": "ephemeral", "ttl": ttl}
        last = messages[-1]
        content = getattr(last, "content", None)
        if isinstance(content, str):
            new_content: list = [{"type": "text", "text": content, "cache_control": marker}]
        elif isinstance(content, list) and content:
            new_content = list(content)
            tail = new_content[-1]
            if isinstance(tail, dict):
                tail = dict(tail)
                tail["cache_control"] = marker
                new_content[-1] = tail
            else:
                new_content.append({"type": "text", "text": str(tail), "cache_control": marker})
        else:
            return messages

        try:
            patched = last.model_copy(update={"content": new_content})
        except Exception:  # noqa: BLE001
            patched = type(last)(content=new_content)
        return [*messages[:-1], patched]

    def _invoke_search(self, args: dict) -> str:
        try:
            out = self._search_tool.invoke(args)
        except Exception as exc:  # noqa: BLE001
            logger.warning("research: search tool raised %s; agent will continue", exc)
            return f"Search error: {exc}"
        return self._stringify(out)

    def _compress(self, text: str, query: str) -> str:
        if not text or _estimate_tokens(text) < self._min_compress_tokens:
            return text
        client = self._engine._compresr_client
        try:
            resp = client.compress(
                context=text,
                query=query,
                compression_model_name=self._compression_model,
            )
            data = getattr(resp, "data", None)
            compressed = getattr(data, "compressed_context", None) if data is not None else None
            return compressed if isinstance(compressed, str) and compressed else text
        except Exception as exc:  # noqa: BLE001
            logger.debug("research: compression failed (%s); using raw snippet", exc)
            return text

    def _maybe_truncate(self, messages: list) -> None:
        if not self._max_context_tokens:
            return
        total = sum(_estimate_tokens(self._stringify(getattr(m, "content", ""))) for m in messages)
        if total <= self._max_context_tokens:
            return
        guard = 0
        while total > self._max_context_tokens and len(messages) >= 4 and guard < 4:
            del messages[2:4]
            total = sum(
                _estimate_tokens(self._stringify(getattr(m, "content", ""))) for m in messages
            )
            guard += 1

    def _aggregate_usage(
        self,
        ai_messages: Sequence[Any],
        *,
        search_calls: int = 0,
    ) -> ResearchUsage:
        from compresr.agents.engine import _aggregate_usage

        agg = _aggregate_usage(list(ai_messages))
        return ResearchUsage(
            input_tokens=int(agg.get("input_tokens", 0) or 0),
            output_tokens=int(agg.get("output_tokens", 0) or 0),
            cache_read_tokens=int(agg.get("cache_read_input_tokens", 0) or 0),
            cache_creation_tokens=int(agg.get("cache_creation_input_tokens", 0) or 0),
            calls=len(ai_messages),
            search_calls=int(search_calls or 0),
        )

    def _collect_citations(
        self, parsed_urls: Sequence[str], trajectory: Sequence[Step]
    ) -> list[Citation]:
        urls: list[str] = []
        for u in parsed_urls:
            if u not in urls:
                urls.append(u)
        for step in trajectory:
            if step.type != "tool_result" or not step.text:
                continue
            for url in re.findall(r"https?://[^\s,;<>\"')]+", step.text):
                if url not in urls:
                    urls.append(url)
        return [Citation(url=u) for u in urls]

    @staticmethod
    def _stringify(content: Any) -> str:
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            parts = []
            for block in content:
                if isinstance(block, dict):
                    text = block.get("text")
                    if isinstance(text, str):
                        parts.append(text)
                elif isinstance(block, str):
                    parts.append(block)
            return "".join(parts)
        return str(content) if content is not None else ""
